src/protocols/secure_channel/CASEServer.cpp

```diff
@@ -73,7 +73,8 @@ CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
     }
     ReturnErrorCodeIf(admin == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
 
-    ReturnErrorOnFailure(admin->GetCredentials(mCredentials, mCertificates, mRootKeyId));
+    uint8_t credentialsIndex;
+    ReturnErrorOnFailure(admin->GetCredentials(mCredentials, mCertificates, mRootKeyId, credentialsIndex));
```

Comment:
If this returns on failure, the `mAdmins->LoadFromStorage(mAdminId)` causes memory to remain cached and not released. How is it cleaned-up?

Reply:
It should be handled by DeviceController::Shutdown() method

Reply:
The credentials are only ever needed during session establishment. They should be loaded at time of use in a scope context, such that they are immediately released when no longer needed. Sigma1 and Sigma 3 require access (Sigma1 only to the root public key of current target fabric), so does Sigma2, on the way back. There is no need at all to maintain the credentials in memory forever outside of CASE, therefore `DeviceController::Shutdown()` is too late to release.

Reply:
This comment was not addressed with a response.

Reply:
The credentials are released right after a successful or unsuccessful CASE Session Establishment. See CASEServer::Cleanup method. There is also issue #8342 to update this method in a future PR.